### PR TITLE
Improve UX for action processor

### PR DESCRIPTION
**Pull request by tushar-composio**
Adds a `processors` field on the `execute_action` and `get_tools` methods, and deprecates the use of `processors` field in the `ComposioToolSet` constructor.

----

> [!IMPORTANT]
> Adds `processors` parameter to `execute_action` and `get_tools`, deprecating its use in `ComposioToolSet` constructor, with tests for new functionality.
> 
>   - **Behavior**:
>     - Adds `processors` parameter to `execute_action` and `get_tools` methods in `ComposioToolSet` and derived classes.
>     - Deprecates `processors` parameter in `ComposioToolSet` constructor.
>   - **Methods**:
>     - Updates `execute_action` in `toolset.py` to merge processors if provided.
>     - Updates `get_tools` in `toolset.py` and derived classes to handle processors.
>   - **Deprecation**:
>     - Adds deprecation warning for `processors` in `ComposioToolSet` constructor.
>   - **Tests**:
>     - Adds tests in `test_toolset.py` for new `processors` handling in `execute_action` and `get_tools` methods.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=SamparkAI%2Fcomposio&utm_source=github&utm_medium=referral)<sup> for fe574c631348929e6980ae8a07683bcb620cbb50. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: python/composio/tools/toolset.py [hunk 1]
```diff
@@ -628,6 +646,22 @@ def _process_schema_properties(self, action: Action, properties: t.Dict) -> t.Di
             type_="schema",
         )
 
+    def _merge_processors(self, processors: ProcessorsType) -> None:
+        for processor_type in self._processors.keys():
+            if processor_type in processors:
+                processor_type = t.cast(
+                    te.Literal["pre", "post", "schema"], processor_type
```
Comment (angrybayblade) [inline python/composio/tools/toolset.py:653]: Can extract `te.Literal["pre", "post", "schema"]` as `ProcessorType`

File: python/tox.ini [hunk 2]
```diff
@@ -124,14 +124,15 @@ commands =
 
     ; TODO: Extract plugin tests separately
     ; Installing separately because of the dependency conflicts
+    pip install plugins/langchain
```
[the remark this answers is not in the document]
Comment (angrybayblade) [reply python/tox.ini:127]: Reason is, tox installs the current top-level package (composio-core) in this case before running the tests and running `pip install plugins/langchain` might overwrite the composio-core installation which might lead to tests breaking.

**Review by angrybayblade — APPROVED** [review]

File: python/composio/tools/toolset.py [hunk 3]
```diff
@@ -67,22 +67,28 @@
 P = te.ParamSpec("P")
 
 _KeyType = t.Union[AppType, ActionType]
-_ProcessorType = t.Callable[[t.Dict], t.Dict]
+_CallableType = t.Callable[[t.Dict], t.Dict]
```
Comment (shreysingla11) [inline python/composio/tools/toolset.py:70]: Consider using a more specific type for `_CallableType`. Instead of `t.Callable[[t.Dict], t.Dict]`, you could use `t.Callable[[t.Dict[str, t.Any]], t.Dict[str, t.Any]]` to be more precise about the input and output types.

File: python/composio/tools/toolset.py [hunk 4]
```diff
@@ -251,18 +257,24 @@ def _limit_file_search_response(response: t.Dict) -> t.Dict:
             self._api_key = None
             self.logger.debug("`api_key` is not set when initializing toolset.")
 
-        self._processors = (
-            processors
-            if processors is not None
-            else {"post": {}, "pre": {}, "schema": {}}
-        )
+        if processors is not None:
+            warnings.warn(
```
Comment (shreysingla11) [inline python/composio/tools/toolset.py:261]: The warning message here could be more specific. Instead of just mentioning 'processors', it could say "Setting 'processors' directly on the ToolSet instance is deprecated..."

File: python/composio/tools/toolset.py [hunk 5]
```diff
@@ -628,6 +650,22 @@ def _process_schema_properties(self, action: Action, properties: t.Dict) -> t.Di
             type_="schema",
         )
 
+    def _merge_processors(self, processors: ProcessorsType) -> None:
```
Comment (shreysingla11) [inline python/composio/tools/toolset.py:653]: Consider adding a type hint for the `processors` parameter in the `_merge_processors` method signature.

File: python/tests/test_tools/test_toolset.py [hunk 6]
```diff
@@ -152,3 +154,89 @@ def test_api_key_missing() -> None:
...
+
+    # Happy case
+    toolset.execute_action(action=Action.GMAIL_FETCH_EMAILS, params={})
+    assert preprocessor_called
+    assert postprocessor_called
+
+    # Improperly defined processors
+    preprocessor_called = postprocessor_called = False
+
+    def weird_postprocessor(reponse: dict) -> None:
```
Comment (shreysingla11) [inline python/tests/test_tools/test_toolset.py:190]: The variable name `reponse` in the `weird_postprocessor` function is misspelled. It should be `response`.

**Comment by shreysingla11** [comment]
I've completed the review of this pull request. Overall, the changes look good and improve the functionality of the toolset. Here's a summary of my findings:

1. The code introduces a new `processors` parameter in various methods, which enhances flexibility.
2. There are some minor improvements that could be made in type annotations and variable naming.
3. The deprecation warning for setting processors directly on the ToolSet instance is a good practice.

In terms of code quality, I'd rate this PR as 8/10. The changes are well-structured and improve the existing codebase. The addition of new tests is particularly commendable. There are just a few minor points that could be addressed to make it even better.

Great job on this PR!

File: python/composio/tools/toolset.py (hunk 5 above)
Comment (shreysingla11) [inline python/composio/tools/toolset.py:653]: Consider adding a type hint for the `processors` parameter in the `_merge_processors` method to improve code readability and maintainability.

File: python/tests/test_tools/test_toolset.py (hunk 6 above)
Comment (shreysingla11) [inline python/tests/test_tools/test_toolset.py:190]: The `weird_postprocessor` function in the test case doesn't follow the expected signature for a post-processor. It should return a dict, but it's currently returning None. Consider updating the function to return the modified response.

**Comment by shreysingla11** [comment]
I've completed the review of the pull request. Overall, the changes look good and improve the functionality of the toolset by adding support for processors in various methods. Here's a summary of my observations:

1. The implementation of processors in the `ComposioToolSet` class and its subclasses is consistent and well-structured.
2. The deprecation warning for setting processors in the constructor is a good approach for maintaining backward compatibility while encouraging the use of the new method.
3. The test cases cover various scenarios, including the happy path and edge cases with improperly defined processors.

There are a couple of minor suggestions for improvement:

1. Consider adding a type hint for the `processors` parameter in the `_merge_processors` method to improve code readability.
2. In the test case for the `weird_postprocessor`, it might be helpful to add a comment explaining that the function intentionally doesn't follow the expected signature to test the warning system.

Overall, the code quality is good, and the changes seem to enhance the flexibility and usability of the toolset. Good job!

File: python/composio/tools/toolset.py [hunk 7]
```diff
@@ -67,22 +67,28 @@
...
 
 _KeyType = t.Union[AppType, ActionType]
-_ProcessorType = t.Callable[[t.Dict], t.Dict]
+_CallableType = t.Callable[[t.Dict], t.Dict]
 
 MetadataType = t.Dict[_KeyType, t.Dict]
 ParamType = t.TypeVar("ParamType")
 
+# Enable deprecation warnings
+warnings.simplefilter("always", DeprecationWarning)
```
Comment (shreysingla11) [inline python/composio/tools/toolset.py:76]: Consider adding a type hint for the `warnings` module to improve code readability and maintainability.

File: python/composio/tools/toolset.py [hunk 8]
```diff
@@ -67,22 +67,28 @@
...
+
+
+ProcessorType = te.Literal["pre", "post", "schema"]
```
Comment (shreysingla11) [inline python/composio/tools/toolset.py:79]: The `ProcessorType` type alias is defined but not used in this file. Consider removing it if it's not needed, or use it to improve type checking in the relevant parts of the code.

File: python/composio/tools/toolset.py (hunk 5 above)
Comment (shreysingla11) [inline python/composio/tools/toolset.py:653]: The `_merge_processors` method modifies the `self._processors` dictionary in-place. Consider returning a new dictionary instead of modifying the existing one to maintain immutability and prevent potential side effects.

File: python/tests/test_tools/test_toolset.py (hunk 6 above)
Comment (shreysingla11) [inline python/tests/test_tools/test_toolset.py:190]: The `weird_postprocessor` function is defined but not used in the test. Consider removing it or using it in a test case to ensure it behaves as expected.

**Comment by shreysingla11** [comment]
I've completed the review of the PR. Here's a summary of the changes and my observations:

1. The PR introduces a new `processors` parameter to the `get_tools()` and `execute_action()` methods, allowing for more flexible pre and post-processing of requests and responses.

2. A deprecation warning has been added for setting processors on the ToolSet initialization, encouraging users to use the new method-level processors instead.

3. The changes are well-tested, with new test cases added to cover the new functionality.

4. The code quality is generally good, with clear naming conventions and logical structure.

Some minor suggestions for improvement:
- Consider adding type hints for imported modules like `warnings`.
- The `ProcessorType` type alias is defined but not used; consider removing it or using it more extensively.
- The `_merge_processors` method modifies state in-place; consider a more functional approach.
- In the test file, there's an unused `weird_postprocessor` function that could be removed or utilized in a test case.

Overall, this PR seems to improve the flexibility and usability of the toolset. The changes are well-implemented and thoroughly tested. With a few minor adjustments, this PR would be ready for merging.

File: python/composio/tools/toolset.py (hunk 5 above)
Comment (shreysingla11) [inline python/composio/tools/toolset.py:653]: In the `_merge_processors` method, consider adding a type check for the `processors` parameter to ensure it's of type `ProcessorsType`. This will help catch potential errors early and improve type safety.

File: python/composio/tools/toolset.py [hunk 9]
```diff
@@ -628,6 +650,22 @@ def _process_schema_properties(self, action: Action, properties: t.Dict) -> t.Di
...
+                existing_processors = self._processors[processor_type]
+            else:
+                existing_processors = {}
+                self._processors[processor_type] = existing_processors
+
+            existing_processors.update(new_processors)
+
     @_record_action_if_available
     def execute_action(
         self,
```
Comment (shreysingla11) [inline python/composio/tools/toolset.py:671]: Consider adding type hints for the `processors` parameter in the `execute_action` method. This will improve type checking and make the method signature more consistent with other parts of the codebase.

File: python/composio/tools/toolset.py (hunk 5 above)
Comment (shreysingla11) [inline python/composio/tools/toolset.py:653]: Consider adding a docstring for the new `_merge_processors` method to explain its purpose and usage. This will help other developers understand the method's functionality and how to use it correctly.

File: python/composio/tools/toolset.py [hunk 10]
```diff
@@ -593,6 +608,13 @@ def _process(
                 f" through: {processor.__name__}"
             )
             data = processor(data)
+            # Users may not respect our type annotations and return something that isn't a dict.
```
Comment (shreysingla11) [inline python/composio/tools/toolset.py:611]: Consider adding more comprehensive error handling for cases where processors might fail or return unexpected types. While you've added a warning for unexpected return types, it might be beneficial to have a more robust error handling strategy, possibly including custom exceptions or fallback behavior.

**Comment by shreysingla11** [comment]
Overall, this pull request introduces valuable improvements to the processor functionality in the ComposioToolSet. The addition of the `ProcessorsType`, the `_merge_processors` method, and the optional `processors` parameter in `execute_action` enhances the flexibility of the toolset.

Here's a summary of the changes and some suggestions for further improvement:

1. The new processor functionality allows for more flexible configuration at different levels (toolset initialization, `execute_action`, and `get_tools`).
2. The deprecation warning for the old method of setting processors is a good practice for maintaining backwards compatibility.
3. The updates to all plugin files to include the new `processors` parameter in their `get_tools` method ensures consistency across the codebase.
4. The new tests in `test_toolset.py` provide good coverage for the new functionality.

Suggestions for further improvement:

1. Consider adding type hints for the `processors` parameter in the `execute_action` method for consistency and improved type checking.
2. Add a docstring for the new `_merge_processors` method to explain its purpose and usage.
3. Consider implementing more comprehensive error handling for cases where processors might fail or return unexpected types.
4. Provide clear documentation and examples in the README or documentation files on how to use the new processors functionality.
5. Consider adding performance benchmarks to ensure that the addition of processors doesn't significantly impact the toolset's performance.
6. Look into a more comprehensive solution for managing plugin dependencies in the future, as the current approach of installing langchain separately might not scale well for multiple plugins.

Overall, this is a solid improvement to the codebase that enhances its flexibility and maintainability. With a few minor additions to documentation and error handling, it will be even more robust and user-friendly.
